experimental/Schemes/Resolution_structure.jl

```diff
@@ -0,0 +1,738 @@
+  j = 0
+  while (( !is_one(I) && b == 0 ) || ( j < b ))
+    push!(Delta_list, I)
+    inc = Oscar.CoveredClosedEmbedding(scheme(I),I)
+    I = Oscar._delta_ideal_for_order(inc)
+    j = j + 1
+  end
+
+  return Delta_list
+end
+
+function _delta_ideal_for_order(inc::CoveredClosedEmbedding, Cov::Covering, 
+       ambient_param_data::IdDict{<:AbsAffineScheme,
+                                 <: Tuple{Vector{Int64},Vector{Int64},<:RingElem}};
+       check::Bool=true)
+
+  W = codomain(inc)                                
+  @check is_smooth(W) "codomain of embedding needs to be smooth"
+#  @check is_equidimensional(W) "codomain of embedding needs to be equidimensional"
+  I_X = small_generating_set(image_ideal(inc))         # ideal sheaf describing X on W
+
+  Delta_dict = IdDict{AbsAffineScheme,Ideal}()
+  for U in Cov
+    I = I_X(U)
+    if is_one(I)
+      Delta_dict[U] = I
+      continue
+    end
+
+    amb_row,amb_col,h = ambient_param_data[U]
+    mod_gens = lifted_numerator.(gens(modulus(OO(U))))
+    R = ambient_coordinate_ring(U)
+    JM = jacobian_matrix(R, mod_gens)
+    if length(amb_col) < length(mod_gens)
+      JM_essential = JM[:, amb_col]
+    else
+      JM_essential = JM
+    end
+    submat_for_minor = JM[amb_row, amb_col]
+    Ainv, h2 = pseudo_inv(submat_for_minor)
+    h == h2 || error("inconsistent input data")
+    JM_essential = JM_essential * Ainv
+    I_gens = lifted_numerator.(gens(I))
+    JI = jacobian_matrix(I_gens)
+    result_mat = h*JI
+    for i in 1:length(amb_col)
+      for j in 1:ncols(result_mat)
+        result_mat[:,j] = result_mat[:,j] - [JM_essential[k,i] * JI[amb_row[i],j] for k in 1:nrows(JM_essential)]
+      end
+    end
+    Ivec = copy(gens(I))
+    append!(Ivec,[a for a in OO(U).(collect(result_mat))])
+    Delta_dict[U] = ideal(OO(U),Ivec)
+  end
+
+  return small_generating_set(IdealSheaf(W,Delta_dict))
+end
+ 
+########################################################################
+# test for snc                                                         #
+########################################################################
+
+function divisor_intersections_with_X(current_div, I_X)
+  scheme(I_X) == scheme(current_div[1]) || error("underlying schemes do not match")
+  n_max = dim(I_X)
+
+  inter_div_dict = Dict{Vector{Int},Tuple{AbsIdealSheaf,Int}}()
+  old_keys = Vector{Int}[]
+  empty_keys = Vector{Int}[]
+  essential_inter = AbsIdealSheaf[]
+
+# initialization: each divisor + I_X
+  for k in 1:length(current_div)
+    Idiv = ideal_sheaf(current_div[k]) + I_X
+    if !is_one(Idiv)
+      inter_div_dict[[k]] = (Idiv,0)
+      push!(old_keys, [k])
+    end
+  end
+  new_keys = copy(empty_keys)
+
+# add intersections
+  while !is_empty(old_keys)
+    old_keyvec = popfirst!(old_keys)         # this is the intersection to which we add a new divisor
+    for i in (old_keyvec[end]+1):length(current_div)
+      copykey = copy(old_keyvec)
+      if haskey(inter_div_dict, [i])
```

Comment:
No need to make a copy if you then don't need it
```suggestion
      if haskey(inter_div_dict, [i])
        copykey = copy(old_keyvec)
```

Personally I also prefer not to nest logic to deeply, as it makes it harder to read, so I probably would do
```suggestion
      haskey(inter_div_dict, [i]) || continue
      copykey = copy(old_keyvec)
```
and then adjust the indentation of the following code.

And then do something similar for `if !is_one(Idiv)` below, changing it to `!is_one(Idiv) || continue` or `is_one(Idiv) && continue` (which one is a matter of taste)

Reply:
done